File: r/polynomial_regression.R
# ==================================================
# 📈 Polynomial Regression: Salary vs Level in R
# Author: Chukwuka Chijioke Jerry
# ==================================================

# 1. Load Required Libraries
library(ggplot2)

# 2. Import Dataset
dataset <- read.csv('../data/Position_Salaries.csv')
dataset <- dataset[ , -1]  # Remove 'Position' column if not needed

# 3. Fit Linear Regression Model
lin_reg <- lm(formula = Salary ~ Level, data = dataset)
summary(lin_reg)

# 4. Fit Polynomial Regression Model (Degree 4)
dataset$Level2 <- dataset$Level^2
dataset$Level3 <- dataset$Level^3
dataset$Level4 <- dataset$Level^4
poly_reg <- lm(formula = Salary ~ Level + Level2 + Level3 + Level4, data = dataset)
summary(poly_reg)

# 5. Visualize Linear Regression Results
ggplot(dataset, aes(x = Level, y = Salary)) +
  geom_point(color = 'blue', size = 2) +
  geom_line(aes(y = predict(lin_reg, newdata = dataset)), color = 'red', linewidth = 1.2) +
  labs(title = 'Salary vs Level (Linear Regression)',
       x = 'Level', y = 'Salary') +
  theme_minimal()

# 6. Visualize Polynomial Regression Results
ggplot(dataset, aes(x = Level, y = Salary)) +
  geom_point(color = 'blue', size = 2) +
  geom_line(aes(y = predict(poly_reg, newdata = dataset)), color = 'red', linewidth = 1.2) +
  labs(title = 'Salary vs Level (Polynomial Regression)',
       x = 'Level', y = 'Salary') +
  theme_minimal()

# 7. Predict Salary for Level 6.5 using Linear Regression
linear_pred <- predict(lin_reg, newdata = data.frame(Level = 6.5))
cat("Linear Regression Prediction for Level 6.5: $", round(linear_pred, 2), "\n")

# 8. Predict Salary for Level 6.5 using Polynomial Regression
poly_pred <- predict(poly_reg, newdata = data.frame(
  Level = 6.5,
  Level2 = 6.5^2,
  Level3 = 6.5^3,
  Level4 = 6.5^4
))
cat("Polynomial Regression Prediction for Level 6.5: $", round(poly_pred, 2), "\n")

# 9. Visualize Polynomial Regression with Higher Resolution
X_grid <- seq(min(dataset$Level), max(dataset$Level), by = 0.01)
high_res_data <- data.frame(
  Level = X_grid,
  Level2 = X_grid^2,
  Level3 = X_grid^3,
  Level4 = X_grid^4
)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = 'blue', size = 2) +
  geom_line(aes(x = X_grid, y = predict(poly_reg, newdata = high_res_data)),
            color = 'red', linewidth = 1.2) +
  labs(title = 'Polynomial Regression-R (High Resolution)',
       x = 'Level', y = 'Salary') +
  theme_minimal()

File: python/polynomial_regression.py
# ==================================================
# 📈 Polynomial Regression: Salary vs Level
# Language: Python
# Author: Chukwuka Chijioke Jerry
# ==================================================

# 📁 1. Import Required Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score

# 📁 2. Load Dataset
dataset = pd.read_csv('../Data/Position_Salaries.csv')
X = dataset.iloc[:, [1]].values  # Extract Level column as feature matrix
y = dataset.iloc[:, -1].values   # Extract Salary as target

# 🧠 3. Fit Linear Regression Model
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# 🧠 4. Fit Polynomial Regression Model
poly_feat = PolynomialFeatures(degree=3)
X_poly = poly_feat.fit_transform(X)
poly_reg = LinearRegression()
poly_reg.fit(X_poly, y)

# 📈 5. Visualize Linear Regression Fit
plt.scatter(X, y, color='blue', label='Actual Data')
plt.plot(X, lin_reg.predict(X), color='red', label='Linear Fit')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.title('Salary vs Level (Linear Regression)')
plt.grid(True)
plt.legend()
plt.tight_layout()
plt.show()

# 📈 6. Visualize Polynomial Regression Fit
plt.scatter(X, y, color='blue', label='Actual Data')
plt.plot(X, poly_reg.predict(X_poly), color='red', label='Polynomial Fit')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.title('Salary vs Level (Polynomial Regression)')
plt.grid(True)
plt.legend()
plt.tight_layout()
plt.show()

# 📈 7. High-Resolution Polynomial Curve
X_grid = np.linspace(X.min(), X.max(), num=100).reshape(-1, 1)
X_grid_poly = poly_feat.transform(X_grid)
plt.scatter(X, y, color='blue', label='Actual Data')
plt.plot(X_grid, poly_reg.predict(X_grid_poly), color='red', label='High-Res Polynomial Fit')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.title('Salary vs Level- Python (High-Resolution Polynomial Regression)')
plt.grid(True)
plt.legend()
plt.tight_layout()
plt.show()

# 🔍 8. Make Predictions
level = 6.5
linear_pred = lin_reg.predict([[level]])[0]
poly_pred = poly_reg.predict(poly_feat.transform([[level]]))[0]

print(f"Linear Prediction for Level {level}: ${linear_pred:,.2f}")
print(f"Polynomial Prediction for Level {level}: ${poly_pred:,.2f}")

# 📊 9. Evaluate Models Using R² Score
r2_linear = r2_score(y, lin_reg.predict(X))
r2_poly = r2_score(y, poly_reg.predict(X_poly))

print(f"Linear Regression R² Score: {r2_linear:.4f}")
print(f"Polynomial Regression R² Score: {r2_poly:.4f}")
